### PR TITLE
`struct DisjointMut`: `unsafe impl Send + Sync`

### PR DESCRIPTION
**Pull request by kkysen**
Does this reasoning make sense?

In particular, I'm unsure about the provenanceless thing and data races not leading to memory unsafety.  Is this worth it to say?  Is it worth it to add an `unsafe trait Provenanceless` to ensure such a thing at a more detailed level?  I do think it's an important point that data races won't lead to memory unsafety, though.

### CODE REVIEW

File: src/disjoint_mut.rs [hunk 1]
```diff
@@ -48,6 +48,40 @@ pub struct DisjointMut<T: ?Sized + AsMutPtr> {
...
+/// that's a different story and justified below for the `unsafe impl Sync`.
+unsafe impl<T: ?Sized + AsMutPtr + Send> Send for DisjointMut<T> {}
+
+/// SAFETY: [`DisjointMut`] only provides disjoint mutable access
+/// to `T`'s elements through a shared `&`[`DisjointMut`]`<T>` reference.
+/// Thus, sharing/[`Send`]ing a `&`[`DisjointMut`]`<T>` across threads is safe.
+///
+/// More precisely, `&`[`Self`] has the two core methods of
+/// [`Self::index`] and [`Self::index_mut`]
+/// that provide disjoint immutable and mutable access to `T`'s elements.
```
Comment (rinon) [inline src/disjoint_mut.rs:64]: I'd like to make it very explicit that the safety of this Sync impl depends on correct usage of `index_mut` which is not marked as unsafe and requires safe usage. That's mentioned throughout this but can you boil it down to a sentence at the top in addition to the long explanation?

File: src/disjoint_mut.rs [hunk 2]
```diff
@@ -48,6 +48,40 @@ pub struct DisjointMut<T: ?Sized + AsMutPtr> {
     inner: UnsafeCell<T>,
 }
 
+/// SAFETY: If `T: `[`Send`], then sending [`DisjointMut`]`<T>` across threads is safe.
+/// There is no non-[`Sync`] state that is left on another thread
+/// when [`DisjointMut`] gets sent to another thread.
+/// When `&`[`DisjointMut`] gets sent to another thread,
+/// that's a different story and justified below for the `unsafe impl Sync`.
```
[the remark this answers is not in the document]
Comment (kkysen) [reply src/disjoint_mut.rs:55]: Okay I'll remove them.

File: src/disjoint_mut.rs [hunk 3]
```diff
@@ -48,6 +48,40 @@ pub struct DisjointMut<T: ?Sized + AsMutPtr> {
...
+/// that provide disjoint immutable and mutable access to `T`'s elements.
+/// This disjointness guarantees that we do not create overlapping `&`s and `&mut`s,
+/// which would be unsound and otherwise possible from multiple threads.
+/// Furthermore, the safety guarantees of [`AsMutPtr::as_mut_ptr`]
+/// ensure that no intermediary `&` or `&mut` references to `T`'s elements are created.
+///
+/// The disjointness guarantee is checked at runtime in debug mode
+/// (i.e. [`#[cfg(debug_assertions)]`]), while in release mode,
+/// disjointness must be manually guaranteed.
+/// This is less than ideal, but crucial for performance.
```
[the remark this answers is not in the document]
Comment (kkysen) [reply src/disjoint_mut.rs:73]: This is important enough that I think we can repeat ourselves a bit.

File: src/disjoint_mut.rs (hunk 1 above)
Comment (kkysen) [reply src/disjoint_mut.rs:64]: How's it now?

**Comment by kkysen** [comment]
Do we need the `T: Send + Sync` bounds in the `unsafe impl Send + Sync` here?  Because as is, `Rav1dPictureDataComponent` is not `Send` or `Sync` (which we need it to be) since it contains a `ptr: NonNull<u8>`.  This `ptr` is not `Sync`, however, as no synchronization is done.

**Comment by rinon** [comment]
I think we need to require it. If that pointer is indeed safe to share we need to mark it as such.

**Review by rinon — APPROVED** [review]

**Comment by kkysen** [comment]
> I think we need to require it. If that pointer is indeed safe to share we need to mark it as such.

It's not safe, though.  That's the problem.

**Comment by kkysen** [comment]
Let me think about it some more.

**Pull request merged**